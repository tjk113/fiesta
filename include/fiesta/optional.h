#pragma once

#ifndef _WIN32
#include <sys/types.h>
#endif
#include <stdint.h>
#include <stddef.h>

/* Optional */

// Optional type wrapping a value of `type` (must be defined using `DEFINE_OPTIONAL`).
#define Optional(type) Optional_##type
// Optional type wrapping a pointer to `type` (must be defined using `DEFINE_OPTIONAL`).
#define OptionalPtr(type) Optional_##type##_Ptr

// Define optional value and pointer types for `type`.
#define DEFINE_OPTIONAL(type) \
    typedef struct {          \
        type val;             \
        bool is_none;         \
    } Optional(type);         \
    typedef struct {          \
        type* val;            \
        bool is_none;         \
    } OptionalPtr(type);

//\ Special case for void*, because void
//\ is always a pointer, never a value.
typedef struct {
    void* val;
    bool is_none;
} OptionalPtr(void);

DEFINE_OPTIONAL(bool)
DEFINE_OPTIONAL(short)
DEFINE_OPTIONAL(char)
DEFINE_OPTIONAL(long)
DEFINE_OPTIONAL(int)
DEFINE_OPTIONAL(size_t)
DEFINE_OPTIONAL(ssize_t)
DEFINE_OPTIONAL(int8_t)
DEFINE_OPTIONAL(uint8_t)
DEFINE_OPTIONAL(int16_t)
DEFINE_OPTIONAL(uint16_t)
DEFINE_OPTIONAL(int32_t)
DEFINE_OPTIONAL(uint32_t)
DEFINE_OPTIONAL(int64_t)
DEFINE_OPTIONAL(uint64_t)
DEFINE_OPTIONAL(float)
DEFINE_OPTIONAL(double)

// An optional value of None.
#define None(type) (Optional(type)){.is_none = true}
// An optional pointer to None.
#define NonePtr(type) (OptionalPtr(type)){.is_none = true}
// An optional value of `type`.
#define Some(type,value) (Optional(type)){.val = (type)value, .is_none = false}
// An optional pointer to `type`.
#define SomePtr(type,value) (OptionalPtr(type)){.val = (type*)value, .is_none = false}
