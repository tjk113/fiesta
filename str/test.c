#include "fiesta/str.h"

int main(int argc, char* argv[]) {
    // The `STR` macro constructs a str from a null-terminated source
    str normal_str = STR("Hello, Fiesta!");
    str_println(normal_str);

    dynstr dynamic_str = DSTR("Horse!");
    dynstr_println(dynamic_str);
    dynstr_remove(&dynamic_str, 2, 3);
    dynstr_println(dynamic_str);

    str_arr split = str_split(normal_str, ' ');
    str_arr_print(split);

    str_free(normal_str);
    dynstr_free(dynamic_str);
    str_arr_free(split);
    return 0;
}