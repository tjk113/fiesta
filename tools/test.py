import subprocess
from enum import Enum

class Color(Enum):
    Blue = '\033[94m'
    Green = '\033[92m'
    Orange = '\033[93m'
    Red = '\033[91m'
    Reset = '\033[0m'
    Bold = '\033[1m'
    Underline = '\033[4m'

class Result(Enum):
    Passed = "Passed"
    Failed = "Failed"

UTILITIES = ["str", "file"]

def report_test_result(module: str, result: Result):
    result_color = ""
    match result:
        case Result.Passed:
            result_color = Color.Green.value
        case Result.Failed:
            result_color = Color.Red.value

    print(f"Module {module}: {result_color}{result.value}{Color.Reset.value}")

for utility in UTILITIES:
    expected = ""
    with open(f"tests/{utility}/expected.txt", "r", encoding="utf-8") as file:
        expected = file.read()

    actual = subprocess.run(
        ["make", "tests"],
        stdout=subprocess.PIPE
    )

    if actual.stdout != expected:
        report_test_result(utility, Result.Failed)
    else:
        report_test_result(utility, Result.Passed)