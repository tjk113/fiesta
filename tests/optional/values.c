#include "optional.h"
#include "test.h"

typedef struct {
    int num_somethings;
    bool is_this_true;
} MyType;

DEFINE_OPTIONAL(MyType)

int main(int argc, char* argv[]) {
    Optional(int) maybe_num = None(int);
    if (maybe_num.is_none)
        puts("none");

    maybe_num = Some(int, 5);
    if (!maybe_num.is_none)
        printf("%d\n", maybe_num.val);

    Optional(MyType) maybe_my_type =
        Some(MyType, ((MyType){.num_somethings = 0, .is_this_true = true}));

    if (!maybe_my_type.is_none)
        printf(
            "%d %s\n",
            maybe_my_type.val.num_somethings,
            maybe_my_type.val.is_this_true ? "true" : "false"
        );

    PASS;
}