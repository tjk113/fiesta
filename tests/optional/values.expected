none
5
0 true
